packages/wrangler/src/__tests__/api/startDevWorker/LocalRuntimeController.test.ts

```diff
@@ -0,0 +1,886 @@
+import events from "node:events";
+import fs, { readFileSync } from "node:fs";
+import net from "node:net";
+import path from "node:path";
+import util from "node:util";
+import { DeferredPromise, Response } from "miniflare";
+import dedent from "ts-dedent";
+import { fetch } from "undici";
+import { assert, describe, expect, it } from "vitest";
+import { WebSocket } from "ws";
+import { LocalRuntimeController } from "../../../api/startDevWorker/LocalRuntimeController";
+import { RuleTypeToModuleType } from "../../../deployment-bundle/module-collection";
+import { teardown, useTmp } from "../../helpers/teardown";
+import { unusable } from "../../helpers/unusable";
+import type {
+	Bundle,
+	File,
+	Module,
+	ReloadCompleteEvent,
+	StartDevWorkerOptions,
+	UrlOriginAndPathnameParts,
+	UrlOriginParts,
+} from "../../../api";
+
+const isWindows = process.platform === "win32";
+function getTextFileContents(file: File<string | Uint8Array>) {
+	if ("contents" in file) {
+		if (typeof file.contents === "string") {
+			return file.contents;
+		}
+		if (file.contents instanceof Buffer) {
+			return file.contents.toString();
+		}
+		return Buffer.from(file.contents).toString();
+	}
+	return readFileSync(file.path, "utf8");
+}
+// // WebAssembly module containing single `func add(i32, i32): i32` export.
+// // Generated using https://webassembly.github.io/wabt/demo/wat2wasm/.
+const WASM_ADD_MODULE = Buffer.from(
+	"AGFzbQEAAAABBwFgAn9/AX8DAgEABwcBA2FkZAAACgkBBwAgACABagsACgRuYW1lAgMBAAA=",
+	"base64"
+);
+
+async function waitForReloadComplete(
+	controller: LocalRuntimeController
+): Promise<ReloadCompleteEvent> {
+	const [event] = await events.once(controller, "reloadComplete");
+	return event;
+}
+
+function joinUrlParts(parts: UrlOriginParts | UrlOriginAndPathnameParts): URL {
+	const pathname = "pathname" in parts ? parts.pathname : "";
+	const spec = `${parts.protocol}//${parts.hostname}:${parts.port}${pathname}`;
+	return new URL(spec);
+}
```

Comment:
This already exists in utils – `urlFromParts`

Reply:
https://github.com/cloudflare/workers-sdk/pull/4450/commits/5dc5115a871e00f821adbe5e0aba3ffd5ad0635d